### DISCUSSION TITLE
Notification on failures?

### DISCUSSION THREAD

**Discussion by DanTulovsky** [argoproj/argo-rollouts#1586] [General]
Hi,

I am setting up: https://argoproj.github.io/argo-rollouts/features/notifications/

It comes with the following pre-defined triggers:

```
  trigger.on-rollout-completed: |
    - send: [rollout-completed]
  trigger.on-rollout-step-completed: |
    - send: [rollout-step-completed]
  trigger.on-rollout-updated: |
    - send: [rollout-updated]
  trigger.on-scaling-replica-set: |
    - send: [scaling-replicaset]
```

Does `on-rollout-completed` include the `failed` case as well?

Thank you
Dan

**Comment by harikrongali** [comment]
@DanTulovsky Currently, only success triggers are added and framework is available to add other remaining especially failures ones.
Users can add custom triggers to manifest files and use. Please contribute back if you have defined more custom triggers. We have it as a lower priority given other priority items.

**Comment by DanTulovsky** [comment]
Would it be possible to provide an example for a failure trigger?

**Comment by jessesuen** [comment]
You are probably interested in `on-rollout-aborted` which translates to the `RolloutAborted` Kubernetes event. We don't have a notification template/trigger for it yet, but it should be easy to add one following the examples here:
https://github.com/argoproj/argo-rollouts/tree/master/manifests/notifications

This would be great to contribute upstream if you come up with a useful template that others can use.

**Comment by DanTulovsky** [comment]
Thanks. I am also interested in "Canary Analysis Failed', how would I get that?

**Comment by DanTulovsky** [comment]
Also, I don't understand, how is a pre-defined trigger like "trigger.on-rollout-step-completed" actually defined?  I understand I can add custom ones, but where do these built-in ones come from?

**Comment by DanTulovsky** [comment]
Oh, i see.. these are just EventReason from `kubectl get events`?  with a "on-" and EventName changed to "event-name" ...

**Comment by DanTulovsky** [comment]
For those following at home, this is for rollout-aborted:

```
  template.rollout-aborted: |
  {{`
    message: Rollout {{.rollout.metadata.namespace}}/{{.rollout.metadata.name}} has been ABORTED.
    email:
      subject: Rollout {{.rollout.metadata.namespace}}/{{.rollout.metadata.name}} has been ABORTED.
    slack:
      attachments: |
          [{
            "title": "{{.rollout.metadata.namespace}}/{{ .rollout.metadata.name}}",
            "color": "#d61317",
            "fields": [
            {
              "title": "Strategy",
              "value": "{{if .rollout.spec.strategy.blueGreen}}BlueGreen{{end}}{{if .rollout.spec.strategy.canary}}Canary{{end}}",
              "short": true
            }
            {{range $index, $c := .rollout.spec.template.spec.containers}}
              {{if not $index}},{{end}}
              {{if $index}},{{end}}
              {
                "title": "{{$c.name}}",
                "value": "{{$c.image}}",
                "short": true
              }
            {{end}}
            ]
          }]
    `}}

  trigger.on-rollout-aborted: |
    - send: [rollout-aborted
```

**Comment by ervinb** [comment]
The list of all built-in events can be found here: https://github.com/argoproj/argo-rollouts/blob/76cfc0e58892ed9759c1a732e243f75aed2fc6bd/utils/conditions/conditions.go#L20-L177

**Reply by martinimartinimartini** [reply]
Hey guys!
I'm configuring notifications for rollouts as well but it doesn't work in "vanilla" way.

I guess I'm suppose to create a when statement for each trigger right?

```yaml
  triggers:
    trigger.on-rollout-completed: |
      - when: rollout.status.conditions[0].reason == 'RolloutCompleted'
        send: 
          - rollout-completed
```

**Reply by ervinb** [reply]
Yes, it's best to configure them explicitly.

The log from your `argo-rollouts` might look something like this:

```sh
Trigger on-rollout-step-completed result: [{[0].2jmj7l5rSw0yVb_vlWAYkK_YBwk  [rollout-step-completed] false}]
```
- `2jmj7l5rSw0yVb_vlWAYkK_YBwk` is the [hash](https://github.com/argoproj/argo-rollouts/blob/e64637171f6555a50c3bdc51820adc3212a2e9b5/utils/record/record.go#L276) of the provided condition (the `when`); in this case the condition was an empty string (default). You can test it [here in this Go Playground](https://go.dev/play/p/_gdnK7YDnLu)
- `[rollout-step-completed]` the event which was recorded
- `false`, the state of the notification: triggered or not

A notification is sent in 2 cases:
1. [Condition is not empty AND the trigger state is `true`](https://github.com/argoproj/argo-rollouts/blob/e64637171f6555a50c3bdc51820adc3212a2e9b5/utils/record/record.go#L289)
2. [Condition was empty (like in the log above), regardless of the state](https://github.com/argoproj/argo-rollouts/blob/e64637171f6555a50c3bdc51820adc3212a2e9b5/utils/record/record.go#L295)
    - there's a caveat to this tough in [notifications-engine](https://github.com/argoproj/notifications-engine/blob/91deed20b9982510f41a61df57fc80cdc1650d4e/pkg/controller/state.go#L51): only the `false => true` state transition sends an actual notification. This can be bypassed/reconfigured by the [oncePer](https://github.com/argoproj/notifications-engine/blob/master/docs/triggers.md#onceper) field

**Reply by IstvanCsVarga** [reply]
This was a tremendous help. Thanks a lot!

**Comment by martinimartinimartini** [comment]
Are you guys been able to create a trigger like this?
```yaml
  trigger.on-rollout-aborted: |
    - send: [rollout-aborted]
```
When I configure it and try to get the trigger list:

```
kubectl exec -it -n argocd argocd-notifications-controller-xxxxxxxxxxx -- /app/argocd-notifications trigger get
failed to parse config: unexpected token EOF (1:1)
```
How to solve it ?
